### PR TITLE
[improve][tests] improved flaky test runs

### PR DESCRIPTION
**Pull request by heesung-sn**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

Fixing flaky tests.

### Modifications

- improved `PulsarFunctionTlsTests` by reordering tearDown() logic
- improved `ManagedLedgerFactoryImpl.shutdown()` by closing cacheEviction threads early
- improved `TestPulsarConnector` memory consumption by removing unnecessary spy()

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change is already covered by existing tests.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API: (yes / **no**)
  - The schema: (yes / **no** / don't know)
  - The default values of configurations: (yes / **no**)
  - The wire protocol: (yes / **no**)
  - The rest endpoints: (yes / **no**)
  - The admin cli options: (yes / **no**)
  - Anything that affects deployment: (yes / **no** / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]
LGTM

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java [hunk 1]
```diff
@@ -1584,7 +1584,7 @@ private void publishAndConsumeMessages(String inputTopic,
         }
 
         for (int i = 0; i < numMessages; i++) {
-            Message<byte[]> msg = consumer.receive(30, TimeUnit.SECONDS);
+            Message<byte[]> msg = consumer.receive();
```
Comment (lhotari) [inline tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java:1587]: Is this change necessary?

**Review by lhotari — APPROVED** [review]

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java (hunk 1 above)
Comment (heesung-sn) [reply tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java:1587]: Because the test env can be really hectic, I think we could see timeouts from time to time. If the response is really delayed, the test will eventually fail. So, yes, I think this will improve the test stability.

Comment (lhotari) [reply tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java:1587]: The test env has reasonable computing resources in CI. Each build job has 2 CPUs and 7GB RAM. I think that it's a bug if receiving a message takes more than 30 seconds. Why would message receiving take more than that?

Comment (heesung-sn) [reply tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java:1587]: I don't have the clear answer here. 

To me, 2 vCPUs might not be enough for the e2e tests, especially when running all pulsar components with dockers. I could be wrong here.

If this timeout issue started happening only recently, then I agree that we have a bug here. Please let me know if we do not want this change. The intention is to make the test more stable for other PRs.

**Review by RobertIndie — APPROVED** [review]

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java (hunk 1 above)
Comment (lhotari) [reply tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java:1587]: > To me, 2 vCPUs might not be enough for the e2e tests, especially when running all pulsar components with dockers. I could be wrong here.

2 vCPUs and 7GB RAM is plenty of computation power & RAM for the tests that we have.

> If this timeout issue started happening only recently, then I agree that we have a bug here. Please let me know if we do not want this change. The intention is to make the test more stable for other PRs.

I am not convinced that we should remove the timeout. The problem must be investigated, and the root cause should be fixed. There must be a bug in production code if 30 seconds isn't sufficient.

**Review by lhotari — CHANGES_REQUESTED** [review]
remove the timeout change from this PR

**Comment by heesung-sn** [comment]
> remove the timeout change from this PR
@lhotari 
I removed it. Please check this again by any chance.

**Review by lhotari — APPROVED** [review]
